# Merging profiles

The **merge** command merges two or more profiles into one, while treating overlapping samples and features in an additive way. This is useful when the analysis includes multiple sets of input files (e.g., multiple sequencing runs).

```bash
woltka tools merge -i input1.biom -i input2.biom -i input3.biom -o output.biom
```

This command is different from QIIME 2's `qiime feature-table merge` command, as the latter can only handle unique sample IDs.

The output file from the merge command is **identical** or nearly identical to the output file generated by merging sequence alignment file prior to running Woltka.

- Very small errors (differring by the count of **1**) could be introduced during the normalization of _multiple assignments_ due to floating point arithmetic issues, which is usually not troublesome. If you do want to avoid this, you may: 1) Use the `--digits` parameter to increase precision level of profiles, which mitigates these errors. 2) Let the upstream alignment step produce unique alignments only (e.g., force one hit per query). 3) Use classification parameters `--rank free`, `--uniq`, `--major`, or `--above` to prevent such small errors ([see details](classify.md#ambiguous-assignment)). 

The **merge** command is generally useful for merging tables. These do not have to be Woltka outputs, nor do they have to be microbiome feature tables. You may supply any feature tables you want to merge.
